File: Python/HouseRobber_III.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return [0, 0]
            left = dfs(node.left)
            right = dfs(node.right)

            robCurrent = node.val + left[1] + right[1]
            skipCurrent = max(left[0], left[1]) + max(right[1], right[0])

            return [robCurrent, skipCurrent]

        robRoot, skipRoot = dfs(root)
        return max(robRoot, skipRoot)

File: JavaScript/HouseRobber_III.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

const rob = function (root) {
  function dfs(node) {
    if (!node) return [0, 0];

    const left = dfs(node.left);
    const right = dfs(node.right);

    const robCurrent = node.val + left[1] + right[1];

    const skipCurrent =
      Math.max(left[0], left[1]) + Math.max(right[0], right[1]);

    return [robCurrent, skipCurrent];
  }
  const [robRoot, skipRoot] = dfs(root);

  return Math.max(robRoot, skipRoot);
};

rob([3, 2, 3, null, 3, null, 1]);
